### ISSUE TITLE
make check error - utf8-2

### ISSUE THREAD

**Issue by mytec33** [gwsw/less#565]
After building 663-beta running `make check` failed with one error:

```
DIFF utf8-2.txt on cmd #150 (E 45)
FAIL: utf8-2.txt (150 steps)
ERR  status 256 from /Users/bill/Source/less-663/lesstest/lesstest  -e  -s '/Users/bill/Source/less-663/lesstest/lt_screen' -t '/Users/bill/Source/less-663/lesstest/lt/utf8-2.txt.lt' '/Users/bill/Source/less-663/less'
RAN  14 tests with 1 errors
make: *** [check] Error 1
```

Steps to build and test:

```
./configure --prefix=/opt/billbrew --with-secure
make -j4
make check
```

System:

```
macOS 14.6.1
ARM
```

I ran `make clean` and reconfigured without `--with-secure`, then  rebuilt and encountered no errors in testing. Repeated the process and added `--with-secure` back and encountered  the error once more.

**Comment by gwsw** [comment]
In general, it would not be expected that a secure build would pass all tests, since by design the secure build makes some commands unavailable, and a test might test one of those commands. However in this case the failure is due to the test invoking the E command, which was not intentional. In 2cfdde53132d7ba0b41c22dabc25620982e1b877 in the post659 branch I have rebuilt that test with the E command invocation removed.

**Comment by avih** [comment]
> I have rebuilt that test with the E command invocation removed.

Are there docs on how to do that? I've read `lesstest/README`, but I don't think it's mentioned specifically how the test files are generated. I.e. I think it mentions how to generate test files, but not how the existing test files are generated.

**Comment by gwsw** [comment]
The README mentions the remaketest command, which is an easy way to rebuild an lt file if there are no changes to the input (the text file and the keypresses). This issue today is actually the first time I've had occasion to rebuild a test while modifying some of the commands. I did it somewhat manually, by running the commands in remaketest individually, and editing the keypress file (the .k file) to remove some keypresses before running maketest.

In other words:
1. ```extract -fv file.lt```
2. ```extract -k file.lt > file.k```
3. Manually edit file.k
4. ```maketest -o file.lt -Okfile.k file.txt```

**Comment by avih** [comment]
> 1. extract -fv file.lt

There's no `file.lt` ...

Maybe there can be a fully automatic script or makefile which creates the "current" tests without manual input?

I.e. that any input keypresses and other files are part of the repo etc, so that someone could run that script and end up witth identical test files as are currently at the repo?

(am I asking a correct question?)

**Comment by mytec33** [comment]
>  In https://github.com/gwsw/less/commit/2cfdde53132d7ba0b41c22dabc25620982e1b877 in the post659 branch I have rebuilt that test with the E command invocation removed.

Confirming all tests now pass. Thank you.

```
./configure --prefix=/opt/billbrew --with-secure
make check
...
make[1]: Nothing to be done for `all'.
RAN  14 tests with 0 errors
bill@MacStudio less % git branch
  master
* post659
```

**Comment by gwsw** [comment]
>There's no file.lt ...

I was just using "file.lt" as a placeholder for whatever lt file you want to rebuild. So to rebuild the utf8-2 file, you'd do 
```
$ cd lesstest
$ ./extract -fv lt/utf8-2.txt.lt
$ ./extract -k lt/utf8-2.txt.lt >utf8-2.txt.k
$ ./maketest -o lt/utf8-2.txt.lt -Okutf8-2.txt.k utf8-2.txt
```

However if you expect to exactly recreate the lt file, you'd need to make sure some environment variables like LESS and LESSCHARSET are set the same way they were set when the file was originally built, and that your screen is the same size as it was when the file was originally built. Even then, the timestamp in the file will be updated so it won't match exactly.

>I.e. that any input keypresses and other files are part of the repo etc, so that someone could run that script and end up witth identical test files as are currently at the repo?
>(am I asking a correct question?)

I'm not sure I understand what your idea is. The .lt file IS the source for the test. Asking for a way to rebuild it sounds kind of like asking how to rebuild main.c. Generally the only time someone would want to rebuild an lt file is if they have modified the less code to make it change its output, causing a test to fail, and they believe that the new output is valid and not a regression. This isn't something the average user or sysadmin would do. Even most changes made to less itself would not require rebuilding any tests (unless less's output was changed). Let me know if I'm misunderstanding what you're asking.

**Comment by avih** [comment]
> I'm not sure I understand what your idea is.
> Let me know if I'm misunderstanding what you're asking.

Neither am I, hence I wasn't sure whether the question is valid. But the reply did clarify few things to me. Thanks.

> The .lt file IS the source for the test.

It's the source file of running the test, but it was generated by (I presume?) manually recording a session with "less" and creating the `.lt` file from that.

So I'm guessing the real input to the current `.lt` files is you recording "less" sessions with various "subjects" actions and input files? I.e. it's not generated using some script, but rather by a human interacting with less in a recorded session.

And now you specifically rebuilt it by extracting the input file and keypresses, editing the keypresses file, and recomposing the `.lt` file?

> Generally the only time someone would want to rebuild an lt file is if they have modified the less code to make it change its output, causing a test to fail, and they believe that the new output is valid and not a regression. This isn't something the average user or sysadmin would do.

Right, but a developer might want to do that, in cases where the expected output/behavior for some input file and key presses has changed.

> Rebuilding an lt file would be done only in conjunction with some substantial change to the less source, and would be considered part of that commit.

I don't think it has to be "substantial". It's probably enough to find a behavior or output bug which happened to get recorded in one of the test files, so once that's fixed, the test would need to be rebuilt, indeed probably at the same commit which fixed the issue.

So I'd think this is something which should be supported/documented (it might already be, I don't know if it is).

**Comment by gwsw** [comment]
>So I'm guessing the real input to the current .lt files is you recording "less" sessions with various "subjects" actions and input files? I.e. it's not generated using some script, but rather by a human interacting with less in a recorded session.

>And now you specifically rebuilt it by extracting the input file and keypresses, editing the keypresses file, and recomposing the .lt file?

Yes, that's all exactly correct.

The remaketest script is intended to rebuild an existing test. Using it is a little fiddly, you have to manually make sure some environment variables etc are correct as I mentioned above. I plan to make this more automated, but rebuilding a test will probably always be a process that takes some care.

Of course, a developer is also free to create their own completely new test if they feel that the existing tests don't adequately cover a change they've made.
